### DISCUSSION TITLE
WPF APIs still needed for WAS projects?

### DISCUSSION THREAD

**Discussion by learntocode66** [microsoft/WindowsAppSDK#3684] [General] [answered]
In my understanding, in development with Windows App SDK, WinRT is used for interactions with Windows system. Then, are those WPF APIs, e.g. those in System.Windows, still needed for WAS projects? Thank you!

Or what's the relationship between [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) and Windows App SDK?

**Comment by DarranRowe** [comment]
To be extremely picky, there is no such thing as a Windows App SDK project.
There is a set of WinUI 3 in Desktop projects. These are projects that sets things up to allow you to use the variant of UWP Xaml that is shipped with the Windows App SDK easier. But this isn't the only way to use the Windows App SDK. If, as an example, I create a C++ console application and then reference the Windows App SDK, I am able to use this functionality in an application even without a visible UI.
So at the most basic level you should think of the Windows App SDK as an extra library that gives you access to some extra features. It isn't a replacement library for the Windows SDK/.NET SDK. The desktop API, either the C++ Windows API, .NET API or even the built in Windows Runtime functionality that is available from desktop applications is still going to be the primary means of interacting with the system.

**Reply by learntocode66** [reply]
Hi @DarranRowe Thank you for your comments!

Sorry for some additional questions:

1. Are [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) part of the so called "Windows SDK"?

_"So at the most basic level you should think of the Windows App SDK as an extra library that gives you access to some extra features."_

2. [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) is part of WPF. So the fact that WPF is to retire does not mean [Windows Desktop APIs] is also to retire. Is this correct?

**Reply by DarranRowe** [reply]
1) The [Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/) is a different entity to what you are referencing. You are referencing the newer .NET framework. The Windows SDK does include the .NET 4 SDK, but not the .NET 5 or newer SDKs. Anyway, in general, when I mention Windows SDK, it is the native C++ side of things, such as [CreateFile](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew), that I mean. This also includes the UWP/Windows Runtime.
2) You have this the wrong way around. WPF is a part of the .NET framework. WPF was introduced as part of .NET 3 all the way back in 2006, where the .NET framework was first released all the way back in 2002. Since WPF is only part of the .NET framework, retiring WPF won't affect the rest of the .NET framework.

**Reply by learntocode66** [reply]
1. _"You are referencing the newer .NET framework."_ 
It seems that [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) is not part of [.NET Framework](https://learn.microsoft.com/en-us/dotnet/api/?view=netframework-4.8.1).
![Image](https://github.com/microsoft/WindowsAppSDK/assets/137069973/34638d7e-ebbe-4e81-8fe1-44b203d51bce)

2. As an example, [Systme.Windows.Clipboard](https://learn.microsoft.com/en-us/dotnet/api/system.windows.clipboard?view=windowsdesktop-7.0) class belongs to [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) and resides in the assembly PresentationCore.dll. So I thought Systme.Windows.Clipboard is part of [Windows Desktop APIs], which also includes APIs like System.Windows.Controls, and should be alias of WPF. Is this wrong?

**Reply by DarranRowe** [reply]
1) It is my terminology that is the issue here. I have a bad habit of calling the newer .NET related stuff .NET Framework. My apologies. However, .NET is still an evolution of the .NET Framework.
If we use the logic that you used in 2., since [System.Windows.Clipboard](https://learn.microsoft.com/en-us/dotnet/api/system.windows.clipboard?view=netframework-4.8.1) belongs to the .NET Framework, does this mean that Windows Desktop API is also part of the .NET Framework? Or the .NET Framework is part of WPF?
The answer to that is no, and the presence of one API under multiple categories just means that it is available in different versions of .NET.

2) You still have this the wrong way around. WPF is part of the .NET API. Hence why you are looking in the .NET API browser and not the WPF API browser.
.NET is much more than WPF, it includes the compilers (C# and F#), the runtime environment and the core system libraries, and these were part of .NET while it was still known as .NET Core. These were also part of .NET Core before desktop development and WPF capabilities were added to .NET Core in .NET Core 3.

**Reply by learntocode66** [reply]
_"WPF capabilities were added to .NET Core in .NET Core 3"_
Are you sure? In my understanding, WPF is NEVER part of new .NET (formerly .NET Core), which is cross-platform and therefore has no APIs for interacting with the Windows system.

https://learn.microsoft.com/en-us/dotnet/desktop/wpf/overview/?view=netdesktop-7.0

> There are two implementations of WPF: The .NET version ... which runs on .NET. ... Even though .NET is a cross-platform technology, WPF only runs on Windows."

And the assembly of [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) is **PresentationCore**.dll, and all the WPF controls are in this category. So I think [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) is simply the alias of WPF. Do you think this is wrong?

![Image](https://github.com/microsoft/WindowsAppSDK/assets/137069973/cf39b3d6-cd70-4180-9292-a383a8730e80)

**Reply by DarranRowe** [reply]
> _"WPF capabilities were added to .NET Core in .NET Core 3"_
> Are you sure? In my understanding, WPF is NEVER part of new .NET (formerly .NET Core), which is cross-platform and therefore has no APIs for interacting with the Windows system.

I am sure.

![Screenshot 2023-06-19 204630](https://github.com/microsoft/WindowsAppSDK/assets/52577874/74ffc3a0-1a01-4f11-a78a-8aa0954f7dc1)

From this screenshot, you can see that PresentationCore.dll is part of the .NET 7 install. In particular, it is the .NET Desktop Runtime 7 package. This Desktop Runtime package is required for this since it provides Windows specific UI libraries. The addition of the desktop package is also documented in the [.NET Core 3 release notes](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0#windows-desktop). There is also a [blog post](https://devblogs.microsoft.com/dotnet/are-your-windows-forms-and-wpf-applications-ready-for-net-core-3-0/) related to this too.
For other platforms, this runtime would be provided by mono or some other provider if it exists.

>And the assembly of [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) is PresentationCore.dll, and all the WPF controls are in this category. So I think [Windows Desktop APIs](https://learn.microsoft.com/en-us/dotnet/api/?view=windowsdesktop-8.0) is simply the alias of WPF. Do you think this is wrong?

First, this is incorrect. [System.IO.FileFormatException](https://learn.microsoft.com/en-us/dotnet/api/system.io.fileformatexception?view=windowsdesktop-8.0) is documented to be in System.IO.Packaging.dll, as an example.
Secondly, [there](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-5#visual-basic-updates) [are](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-7#wpf) [numerous](https://devblogs.microsoft.com/dotnet/wpf-on-dotnet-7/) [sources](https://learn.microsoft.com/en-us/dotnet/core/compatibility/5.0?toc=%2Fdotnet%2Ffundamentals%2Ftoc.json&bc=%2Fdotnet%2Fbreadcrumb%2Ftoc.json#wpf) that place WPF as part of .NET. [There](https://learn.microsoft.com/en-us/dotnet/core/compatibility/8.0?toc=%2Fdotnet%2Ffundamentals%2Ftoc.json&bc=%2Fdotnet%2Fbreadcrumb%2Ftoc.json#windows-forms) [is](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-7#windows-forms) [also](https://devblogs.microsoft.com/dotnet/winforms-enhancements-in-dotnet-7/) [lots](https://learn.microsoft.com/en-us/dotnet/core/compatibility/7.0?toc=%2Fdotnet%2Ffundamentals%2Ftoc.json&bc=%2Fdotnet%2Fbreadcrumb%2Ftoc.json#windows-forms) [of](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-6#windows-forms) [documentation](https://devblogs.microsoft.com/dotnet/whats-new-in-windows-forms-in-net-6-0-preview-5/) [that](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-1#windows-forms) places Windows Forms, which is also part of the [Windows Desktop API](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms?view=windowsdesktop-8.0), as part of .NET too.

**Comment by learntocode66** [comment]
Clipboard as an example:

From: Windows Desktop 7
[System.Windows.Clipboard class](https://learn.microsoft.com/en-us/dotnet/api/system.windows.clipboard?view=windowsdesktop-7.0)

From: WinRT APIs
[Windows.ApplicationModel.DataTransfer.Clipboard class](https://learn.microsoft.com/en-us/uwp/api/windows.applicationmodel.datatransfer.clipboard?view=winrt-22621)

The WinRT version seems to be the replacement of the Windows Desktop 7 version. Is this understanding correct?

**Reply by DarranRowe** [reply, marked as answer]
I honestly don't think it matters. The [.NET source for the clipboard class](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/clipboard.cs) has it using the Windows native OLE clipboard functionality. For example, line 643 has a call to [OleGetClipboard](https://learn.microsoft.com/en-us/windows/win32/api/ole2/nf-ole2-olegetclipboard).
If you dig in far enough, I'm sure that you will find that the Windows Runtime clipboard functionality is also a wrapper around the native OLE clipboard functionality too. So as long as the underlying functionality is supported then you should just use whatever is most comfortable for you.

**Reply by learntocode66** [reply]
Hi @DarranRowe   Thank your so much for your patient clarification of some basic concepts of .NET!
